### DISCUSSION TITLE
Logistic regression - features with zero weight associated with non-zero SHAP values

### DISCUSSION THREAD

**Discussion by inezpereira** [shap/shap#3160] [General]
Hello everyone!

I have a Logistic Regression model which uses Elastic Net regularization (specifically, an [SGDClassifier](https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) with a log loss and an elasticnet penalty). I am using this model because I have a dataset with many relatively many features, and I wanted Elastic Net to do feature selection for me.
I have fitted my model and now calculated the associated SHAP values. I noticed however that, often, even if a feature weight has been set to zero (thanks to the L1 regularization in Elastic Net), I still get a non-zero SHAP value.

My question is: 
Is this a violation of missingness? 
Or is it indeed possible to have non-zero SHAP values even if the weight calculated for this feature is zero?

I've combed several talks by @slundberg, and had a look at [the original paper](https://papers.nips.cc/paper_files/paper/2017/file/8a20a8621978632d76c43dfd28b67767-Paper.pdf), but I am relatively new to SHAP, so your help would be very much appreciated :)

Below this code snippet I am using:
```
X_train_summary = shap.kmeans(X_train, 10)  
explainer = shap.KernelExplainer(model.predict_proba, X_train_summary, feature_names=model.feature_names_in_)
# Calculates the SHAP values - It takes some time
shap_values = explainer.shap_values(X_test)

```
Thanks!

**Comment by detrin** [comment]
Hello @inezpereira, my thinking is as follows.
1. Logistic Regression is linear model so in first place, it doesn't really make that much sense to use shap to explain LogisticRegression as the weights give you perfect explanation what is the boundary between one and second class in binary classification.
2. Regarding your example. If you are using Logistic Regression from scikit-learn, try setting `fit_intercept` to `False`. https://scikit-learn.org/stable/modules/generated/sklearn.linear_model.LogisticRegression.html If you have nonzeroe intercept the feature with zero weight will still play a role in the overall generated prediction.

**Comment by inezpereira** [comment]
Hi there @detrin!
Regarding the use of SHAP with logistic regression: I agree and I moved on to considering the weights, but I still had this more theoretical question.
See [here](https://www.dl.begellhouse.com/download/article/706e4963504bc249/493-514.pdf) under Equation 23: in a linear model, if features are correlated, even if their associated weights are zero, their Shapley values can be non-zero, according to the authors.

**Reply by detrin** [reply]
Please if you have new question open new discussion and mark my previous comment as solved. The article is not open, so when opening new question in discussion make sure that you include the parts of the article you want. Thanks for understanding

**Reply by inezpereira** [reply]
Ah, my apologies! I thought I had checked that the article was open. 
I am uploading a screenshot of the relevant section here, since the paper answers the question
<img width="908" alt="image" src="https://github.com/shap/shap/assets/20027814/01ddee59-dfa1-4966-a6af-57d93713168d">

Thanks for your help!

**Reply by detrin** [reply]
No problem and thanks for the input. Well now it makes more sense for this specific minimal example So now back to your original questions:
> Is this a violation of missingness?

Are you referring to the fact when the weights are set to zero in logistic regression and the shapley values of those features are nonzero? Well, I think as explained in the article it is "okay" according to the definition of shapley values. Depends what you aim to achieve. Personally, when selecting features I prefer removing weak features with shap+boosting model no matter whether the end model is going to be boosting model, or Logistic Regression.

> Or is it indeed possible to have non-zero SHAP values even if the weight calculated for this feature is zero?

As we see in the article, yes it is possible :)

I think that if you have highly correlated features in your Logistic Regression (>=0.7 rule of thumb), then you are in trouble. After removing highly correlated features you could calculate shap values and the whole discussion should not be relevant anymore. To correct my previous statement: It doesn't really make that much sense to use shap to explain LogisticRegression as the weights give you perfect explanation what is the boundary between one and second class in binary classification **when the features are not correlated**.

**Reply by detrin** [reply]
If you are interested in selecting just relevant features, check out https://github.com/ing-bank/probatus, https://github.com/cerlymarco/shap-hypetune

**Reply by inezpereira** [reply]
Awesome, I'll have a look. I have closed this discussion, as my question was answered.

**Reply by detrin** [reply]
I am glad to help :)
